build-logic/src/main/kotlin/ktlint-kotlin-common.gradle.kts

```diff
@@ -5,22 +5,38 @@ plugins {
     kotlin("jvm")
 }
 
+val projectLibs = extensions.getByType<VersionCatalogsExtension>().named("libs")
+val javaCompilationVersion = JavaLanguageVersion.of(projectLibs.findVersion("java-compilation").get().requiredVersion)
+val javaTargetVersion = JavaLanguageVersion.of(projectLibs.findVersion("java-target").get().requiredVersion)
+
 kotlin {
     // All modules, the CLI included, must have an explicit API
     explicitApi()
+    jvmToolchain(jdkVersion = javaCompilationVersion.asInt())
 }
 
+tasks.withType<JavaCompile>().configureEach {
+    options.release.set(javaTargetVersion.asInt())
+}
 tasks.withType<KotlinCompile>().configureEach {
-    compilerOptions {
-        jvmTarget.set(JvmTarget.JVM_1_8)
-    }
+    // Convert Java version (e.g. "1.8" or "11") to Kotlin JvmTarget ("8" resp. "11"
+    compilerOptions.jvmTarget.set(JvmTarget.fromTarget(JavaVersion.toVersion(javaTargetVersion).toString()))
 }
 
-// compileJava task and compileKotlin task jvm target compatibility should be set to the same Java version.
-// For some reason, we fallback from toolchain using, see https://github.com/pinterest/ktlint/pull/1787
-tasks.withType<JavaCompile>().configureEach {
-    sourceCompatibility = JavaVersion.VERSION_1_8.toString()
-    targetCompatibility = JavaVersion.VERSION_1_8.toString()
+val requestedJdkVersion = project.findProperty("testJdkVersion")?.toString()?.toInt()
+// List all non-current Java versions the developers may want to run via IDE click
+setOf(8, 11, 17, requestedJdkVersion).filterNotNull().forEach { version ->
```

Comment:
I tried to explain that all listed values here will show up in the IDE. Most likely this should match the ones CI tests against and be LTS version, but it doesn't have to. The list here is responsible only for tasks that will show up in IDE.

A set of pre-registered tasks are available, but also, one can pass custom `testJdkVersion` (and that's what CI will leverage). `setOf().filterNotNull()` makes sure the build script doesn't register test task for the same Java version twice

Reply:
Cool. Could be written as:
```
setOfNotNull(8, 11, 17, requestedJdkVersion).forEach { version ->
```